multi_source_job_queue.go

```diff
@@ -0,0 +1,65 @@
+package worker
+
+import (
+	"sync"
+	"time"
+
+	gocontext "context"
+
+	"github.com/travis-ci/worker/context"
+)
+
+type MultiSourceJobQueue struct {
+	queues             []JobQueue
+	buildJobChans      []<-chan Job
+	buildJobChansMutex *sync.Mutex
+}
+
+func NewMultiSourceJobQueue(queues ...JobQueue) *MultiSourceJobQueue {
+	return &MultiSourceJobQueue{
+		queues:             queues,
+		buildJobChansMutex: &sync.Mutex{},
+	}
+}
+
+func (tjq *MultiSourceJobQueue) Jobs(ctx gocontext.Context) (outChan <-chan Job, err error) {
+	logger := context.LoggerFromContext(ctx)
+
+	buildJobChan := make(chan Job)
+	outChan = buildJobChan
+
+	go func() {
+		tjq.buildJobChans = []<-chan Job{}
+		for _, queue := range tjq.queues {
+			jc, err := queue.Jobs(ctx)
+			if err != nil {
+				logger.WithField("err", err).Error("failed to get job chan from queue")
+				return
```

Comment:
I believe this logic allows any failure in any `queue.Jobs()` call to break the entire `Jobs()` from succeeding. I might be missing a part where this is handled though.